File: Algorithm/Algorithm_Practice/Codility/CodingTest/ESTSoft/JS/3.js
function solution(A){
    function isZigZag(treeSkyline){
        let direction = '';
        for(let i= 0 ; i< treeSkyline.length - 1 ; i++){
            if(treeSkyline[i] < treeSkyline[i+1]){
                if(direction != ''){
                    if(direction === 'UP'){
                        return false;
                    }
                    else{
                        direction = 'UP';
                    }
                }
                else{
                    direction = 'UP'
                }
            }
            else if(treeSkyline[i] > treeSkyline[i+1]){
                if(direction != ''){
                    if(direction === 'Down'){
                        return false;
                    }
                    else{
                        direction = 'Down';
                    }
                }
                else{
                    direction = 'Down'
                }
            }
            else{
                // Same Height 
                return false;
            }
        }
        return true;
    }

    if(isZigZag(A) == true){
        return 0;
    }
    let cnt = 0;
    for(let i = 0 ; i < A.length ; i++){
        let tmpTreeline = A.slice();
        tmpTreeline.splice(i,1);
        if(isZigZag(tmpTreeline)===true){
            cnt+=1;
        }   
    }
    if (cnt === 0){
        return -1;
    }
    else{
        return cnt;
    }
}

A = [3,4,5,3,7];
B = [1,2,3,4];
C = [1,3,1,2];
console.log(solution(C));

File: Algorithm/Algorithm_Practice/Codility/CodingTest/ESTSoft/Python/3.py

# 인접 나무 같은 높이 금지
# k번째 나무는 
# 불가능 값 return -1

def solution(A):
    def isZigZag(treeSkyLine):
        direction = ''
        for idx in range(len(treeSkyLine)-1):
            if (treeSkyLine[idx+1] < treeSkyLine[idx]):
                if (direction == "Down"):
                    return False
                else:
                    direction = "Down"
            elif(treeSkyLine[idx+1] > treeSkyLine[idx]):
                if(direction == "UP"):
                    return False
                else:
                    direction = "UP"
            else:
                return False
        return True
    if(isZigZag(A) == True):
        return 0 
    else:
        cnt = 0
        for i in range(len(A)):
            CuttedtreeSkyLine = list(A)
            CuttedtreeSkyLine.pop(i)
            if (isZigZag(CuttedtreeSkyLine) == True):
                cnt +=1
        if cnt == 0:
            return -1
        else:
            return cnt

A = [3,4,5,3,7]
B = [1,3,1,2]
C = [1,2,3,4]

print(solution(B))